torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py

```diff
@@ -88,12 +96,68 @@ def aot_dispatch_base_graph(
         fw_only=flat_fn,
     )
 
+    non_strict_exporting_mod = _root_module_for_non_strict_export(flat_fn)
+    if aot_config.is_export and non_strict_exporting_mod is not None:
+        # in non-strict export
+
+        # For any buffer that is assigned, we want to associate it to the final proxy node
+        # that it is assigned to. This node can then be added as a buffer mutation output.
+        assigned_buffers = {}
+
+        def _map_assigned_buffer_to_proxy(_mod, name, buffer):
+            # We intercept buffer assignments by registering this hook.
+            proxy_mode = torch._C._get_dispatch_mode(
+                torch._C._TorchDispatchModeKey.PROXY
+            )
+            if proxy_mode is not None and isinstance(
```

Comment:
Hmm why would proxy mode be none here?

Reply:
@bdhirsh asked the same question. I'm not sure why it can be none, but if I assert it is always non-none then I get various test failures. Let me make a catalog of those as a comment before landing.

Reply:
Here's the list of errors on `fbcode//caffe2/test:test_export` when I assert: https://www.internalfb.com/intern/testinfra/testconsole/testrun/4785074831210563/

Reply:
Here's a partial stack trace for one of the errors related to `cond` (which triggers `torch.compile(fullgraph=True)` which then seems to call `register_buffer` for some kind of setup?):

```
 File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/caffe2/test/export/test_export.py", line 3676, in forward
    return torch.cond(x.shape[0] > 4, true_fn, false_fn, [x])
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_higher_order_ops/cond.py", line 138, in cond
    return torch.compile(cond_op, backend="eager", fullgraph=True)(
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/eval_frame.py", line 410, in _fn
    return fn(*args, **kwargs)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/convert_frame.py", line 966, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state, skip=1)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/convert_frame.py", line 404, in _convert_frame_assert
    return _compile(
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_utils_internal.py", line 279, in wrapper_function
    return StrobelightCompileTimeProfiler.profile_compile_time(
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/caffe2/fb/strobelight/compile_time_profiler.py", line 89, in profile_compile_time
    return func(*args, **kwargs)
  File "/usr/local/fbcode/platform010/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/convert_frame.py", line 691, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/utils.py", line 265, in time_wrapper
    r = func(*args, **kwargs)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/convert_frame.py", line 546, in compile_inner
    out_code = transform_code_object(code, transform)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/bytecode_transformation.py", line 1103, in transform_code_object
    transformations(instructions, code_options)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/convert_frame.py", line 168, in _fn
    return fn(*args, **kwargs)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/convert_frame.py", line 508, in transform
    tracer.run()
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/symbolic_convert.py", line 2198, in run
    super().run()
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/symbolic_convert.py", line 848, in run
    while self.step():
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/symbolic_convert.py", line 758, in step
    self.dispatch_table[inst.opcode](self, inst)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/symbolic_convert.py", line 2334, in RETURN_VALUE
    self._return(inst)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/symbolic_convert.py", line 2319, in _return
    self.output.compile_subgraph(
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/output_graph.py", line 961, in compile_subgraph
    self.compile_and_call_fx_graph(tx, list(reversed(stack_values)), root)
  File "/usr/local/fbcode/platform010/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_dynamo/output_graph.py", line 1142, in compile_and_call_fx_graph
    gm = _make_graph_module(root, self.graph)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/fx/_lazy_graph_module.py", line 58, in _make_graph_module
    return graph_module_cls(*args, **kwargs)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/fx/graph_module.py", line 400, in __init__
    _copy_attr(root, self, node.target)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/fx/graph_module.py", line 234, in _copy_attr
    to_module.register_buffer(field, orig)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/nn/modules/module.py", line 519, in register_buffer
    output = hook(self, name, tensor)
  File "/mnt/xarfuse/uid-8011/1171760a-seed-nspid4026531836_cgpid231623662-ns-4026531840/torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py", line 110, in _map_assigned_buffer_to_proxy
    assert proxy_mode is not None, f"{type(_mod)}.{name} = {type(buffer)}"
```

Reply:
Here's another stack trace for a non-cond test: https://www.internalfb.com/intern/testinfra/diagnostics/4785074831210563.281475113795226.1712792368/

This one for `strict_mode_op`. So a HOP thing?

Reply:
OK I'm going to try to narrow down when the hook is supposed to fire (basically only on the root module of export) to make stronger assertions. 🙏